src/rule_definition_tools.jl

```diff
@@ -366,7 +368,7 @@ function _nondiff_rrule_expr(__source__, primal_sig_parts, primal_invoke)
     tup_expr = tuple_expression(primal_sig_parts)
     primal_name = first(primal_invoke.args)
     pullback_expr = @strip_linenos quote
-        function $(esc(propagator_name(primal_name, :pullback)))(::Any)
+        function $(esc(propagator_name(primal_name, :pullback)))(@nospecialize(_))
```

Comment:
Using
```suggestion
        function $(esc(propagator_name(primal_name, :pullback)))(@nospecialize(::Any))
```
instead leads to problems with the type annotation `::Any` in the `ntuple` expression: the macro would output something like
```julia
        function (ChainRulesCore.ChainRulesCore).rrule(::(Core).Typeof(length), x::AbstractVector...)
            #= REPL[17]:1 =#
            return (length(x...), begin
                        function length_pullback($(Expr(:meta, :nospecialize, :(::Any))))
                            return ChainRulesCore.ntuple(((::var"#179#Any",)->begin
                                            ChainRulesCore.DoesNotExist()
                                        end), 1 + ChainRulesCore.length(x))
                        end
                    end)
        end
```

Reply:
You seem to have a faible for discovering macro hygiene bugs! :laughing: Interestingly, this does not trigger the other bug with incorrectly escaping `_`...

Reply:
https://github.com/JuliaLang/julia/pull/40293